[cocoa] Explicitly set the minimum TLS version to 1.0 when enabled

sysrqb opened the pull request:
#### cad628f39ffbd176042a10c5af8e6cf51b9d5026
<pre>
[cocoa] Explicitly set the minimum TLS version to 1.0 when enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=280145">https://bugs.webkit.org/show_bug.cgi?id=280145</a>
<a href="https://rdar.apple.com/136185611">rdar://136185611</a>

Reviewed by Alex Christensen.

The system default minimum TLS version is changing to TLS 1.2. Continue
allowing older protocol versions if the option is enabled.

Original patch by Guoye Zhang

* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.serialization.in:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::NetworkSessionCocoa::NetworkSessionCocoa):
(WebKit::SessionSet::initializeEphemeralStatelessSessionIfNeeded):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::parameters):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TLSDeprecation.mm:
(TestWebKitAPI::TEST(TLSVersion, DefaultBehavior)):
(TestWebKitAPI::TEST(TLSVersion, NetworkSession)):
(TestWebKitAPI::webViewWithNavigationDelegate):
(TestWebKitAPI::TEST(TLSVersion, LegacySubresources)):
* Tools/TestWebKitAPI/cocoa/HTTPServer.mm:
(TestWebKitAPI::HTTPServer::listenerParameters):

Canonical link: <a href="https://commits.webkit.org/291333@main">https://commits.webkit.org/291333@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/521f8426c3ef8da860eb7e2d7dc3ee77bf874310

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92516 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12061 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1667 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/97500 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43023 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94566 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12338 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20519 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70876 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28322 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95518 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9364 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83777 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51208 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9060 "Found 2 new test failures: imported/w3c/web-platform-tests/svg/struct/scripted/blank.svg imported/w3c/web-platform-tests/webrtc-stats/hardware-capability-stats.https.html (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1414 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42354 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79380 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1356 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99527 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19567 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14441 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79883 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19817 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79658 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79161 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23694 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/989 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/12579 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14775 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19551 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24723 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19238 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22698 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20978 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/TLSDeprecation.mm:
```diff
@@ -191,6 +203,14 @@ - (void)_webView:(WKWebView *)webView didNegotiateModernTLSForURL:(NSURL *)url
+        [configuration setLegacyTLSEnabled:YES];
+        auto dataStore = adoptNS([[WKWebsiteDataStore alloc] _initWithConfiguration:configuration.get()]);
+        auto webView = makeWebViewWith(dataStore.get(), delegate);
+        [webView loadRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:[NSString stringWithFormat:@"https://127.0.0.1:%d/", server.port()]]]];
```
This could just be [webView loadRequest:server.request()]

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/TLSDeprecation.mm:
```diff
@@ -240,7 +272,16 @@ HTTPServer server([&](const Connection&) {
+    RetainPtr<WKWebView> webView;
+    if (!configuration) {
+        auto configuration = adoptNS([[_WKWebsiteDataStoreConfiguration alloc] initNonPersistentConfiguration]);
+        [configuration setLegacyTLSEnabled:YES];
```
Should this be protected by ENABLE(TLS_1_2_DEFAULT_MINIMUM)?

achristensen07 commented on Tools/TestWebKitAPI/cocoa/HTTPServer.mm:
```diff
@@ -133,8 +133,10 @@ static bool shouldDisableTLS(HTTPServer::Protocol protocol)
         sec_protocol_options_set_local_identity(options.get(), identity.get());
-        if (protocol == Protocol::HttpsWithLegacyTLS)
+        if (protocol == Protocol::HttpsWithLegacyTLS) {
+            sec_protocol_options_set_min_tls_protocol_version(options.get(), tls_protocol_version_TLSv10);
```
Protect with ENABLE(TLS_1_2_DEFAULT_MINIMUM)?

guoye-zhang committed:
[cocoa] Explicitly set the minimum TLS version to 1.0 when enabled
https://bugs.webkit.org/show_bug.cgi?id=280145
rdar://136185611

Reviewed by Alex Christensen.

The system default minimum TLS version is changing to TLS 1.2. Continue
allowing older protocol versions if the option is enabled.

Original patch by Guoye Zhang

* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.serialization.in:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::NetworkSessionCocoa::NetworkSessionCocoa):
(WebKit::SessionSet::initializeEphemeralStatelessSessionIfNeeded):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::parameters):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TLSDeprecation.mm:
(TestWebKitAPI::TEST(TLSVersion, DefaultBehavior)):
(TestWebKitAPI::TEST(TLSVersion, NetworkSession)):
(TestWebKitAPI::webViewWithNavigationDelegate):
(TestWebKitAPI::TEST(TLSVersion, LegacySubresources)):
* Tools/TestWebKitAPI/cocoa/HTTPServer.mm:
(TestWebKitAPI::HTTPServer::listenerParameters):

Canonical link: https://commits.webkit.org/291333@main
